contracts/types/YieldSplitter.sol

```diff
@@ -0,0 +1,167 @@
+ */
+abstract contract YieldSplitter {
+    using SafeERC20 for IERC20;
+
+    error YieldSplitter_NotYourDeposit();
+
+    IOHMIndexWrapper public immutable indexWrapper;
+
+    struct DepositInfo {
+        uint256 id;
+        address depositor;
+        uint256 principalAmount; // Total amount of sOhm deposited as principal, 9 decimals.
+        uint256 agnosticAmount; // Total amount deposited priced in gOhm. 18 decimals.
+    }
+
+    uint256 public idCount;
+    mapping(uint256 => DepositInfo) public depositInfo; // depositId -> DepositInfo
+    mapping(address => uint256[]) public depositorIds; // address -> Array of the deposit id's deposited by user
+
+    /**
+        @notice Constructor
+        @param indexWrapper_ Address of contract that will return the sOHM to gOHM index. 
+                             On mainnet this will be sOHM but on other chains can be an oracle wrapper.
+    */
+    constructor(address indexWrapper_) {
+        indexWrapper = IOHMIndexWrapper(indexWrapper_);
+    }
+
+    /**
+        @notice Create a deposit.
+        @param depositor_ Address of depositor
+        @param amount_ Amount in gOhm. 18 decimals.
+    */
+    function _deposit(address depositor_, uint256 amount_) internal returns (uint256 depositId) {
+        depositorIds[depositor_].push(idCount);
+
+        depositInfo[idCount] = DepositInfo({
+            id: idCount,
+            depositor: depositor_,
+            principalAmount: _fromAgnostic(amount_),
+            agnosticAmount: amount_
+        });
+
+        depositId = idCount;
+        idCount++;
+    }
+
+    /**
+        @notice Add more gOhm to the depositor's principal deposit.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOhm to add. 18 decimals.
+    */
+    function _addToDeposit(uint256 id_, uint256 amount_) internal {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        userDeposit.principalAmount += _fromAgnostic(amount_);
+        userDeposit.agnosticAmount += amount_;
+    }
+
+    /**
+        @notice Withdraw part of the principal amount deposited.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOHM to withdraw.
+    */
+    function _withdrawPrincipal(uint256 id_, uint256 amount_) internal {
+        if (depositInfo[id_].depositor != msg.sender) revert YieldSplitter_NotYourDeposit();
+
+        DepositInfo storage userDeposit = depositInfo[id_];
+        userDeposit.principalAmount -= _fromAgnostic(amount_); // Reverts if amount > principal due to underflow
+        userDeposit.agnosticAmount -= amount_;
+    }
+
+    /**
+        @notice Withdraw all of the principal amount deposited.
+        @param id_ Id of the deposit.
+        @return amountWithdrawn : amount of gOHM withdrawn. 18 decimals.
+    */
+    function _withdrawAllPrincipal(uint256 id_) internal returns (uint256 amountWithdrawn) {
+        if (depositInfo[id_].depositor != msg.sender) revert YieldSplitter_NotYourDeposit();
```

Comment:
would this not read the YieldDirectorV2 address as msg.sender? or is it different when inheriting abstracts?

Reply:
that only happens when another contract calls this contract. When you inherit the base contract you just combine and become as one sorta. So this will be the same msg.sender as YieldDirectorV2's msg.sender.

Reply:
very good to know

Reply:
The only concern I have with this check is if we want to do something like allow another contract to withdraw for the user. I see this happening being a thing more for redeeming yield tho so I don't see this happening with withdraw.

Do you think we will have implementations that would want to allow another contract to withdraw on behalf of a user? if so we should probably remove this and make implementation handle this check.

Reply:
I think this contract should handle that case, since it is a base contract to be inherited. I see this as a building block for other systems to be built on, not just the two of them that we have currently.

Reply:
so for example in the Odyssey buy-now-pay-later idea, we would build that directly off of this rather than building something that interacts through say YieldDirector, and thus this check still works?

Reply:
Yea this check will work as long as the depositor will be calling the same contract they deposited to. 

So in the case of an odyssey contract that will build a contract that implements this base contract. As long as depositor is interacting with the contract that implements this contract will be all good. But if theres another contract that will interact with the odyssey contract wants to withdraw on behalf of user would stuff up.